project/src/main/java/com/google/sps/servlets/QueryServlet.java

```diff
@@ -48,17 +53,38 @@ void init(PlacesFetcher inputFetcher) {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    ImmutableList<Place> filteredPlaces;
     try {
-      //TODO(M1): add call to filterer
-      response.setContentType("application/json");
-      response.getWriter().write(new Gson().toJson(
-          Places.randomSort(fetcher.fetch())
-          .stream()
-          .limit(MAX_NUM_PLACES_TO_RECOMMEND)
-          .collect(Collectors.toList())
-      ));
+      UserPreferences userPrefs =
+          UserPreferences.builder()
+              .setMinRating(Float.parseFloat(request.getParameter("rating")))
+              .setMaxPriceLevel(Integer.parseInt(request.getParameter("price")))
+              .setOpenNow(Integer.parseInt(request.getParameter("open")) != 0)
+              .setLocation(getLatLngFromString(request.getParameter("location")))
+              .setCuisines(ImmutableList.copyOf(request.getParameter("cuisines").split(",")))
+              .build();
+      filteredPlaces = Places.filter(
+          fetcher.fetch(userPrefs) /* places */,
+          Integer.parseInt(request.getParameter("rating")) /* min rating */,
+          true /* filter if no website */,
+          true /* filter branches of same place */
+      );
     } catch (Exception e) {
```

Comment:
I think you need to catch specific Exceptions and handle them accordingly. 
For example, the Fetcher throws a `FetcherException`, which means Google maps API had a problem. 
But you might end up with NullpointerException, RuntimeException, or many other exceptions. 
So, first catch what you know might be thrown, and then handle this in a specific way (here, returning an error with the message "Fetching from Google Places API encountered a problem"). 
Then, catch all other exceptions and send the relevant error (which in this case, won't be fetching from google places API).

Reply:
Another option will be to catch only what you intend that will be thrown. Everything else will be returned as server error, which you can then deal with in the Frontend.

Reply:
got it, thanks! Tried to add a more accurate error handling